### DISCUSSION TITLE
Non convergence with modification of PenaltyInterfaceDiffusion

### DISCUSSION THREAD

**Discussion by Chaoyue-Liu** [idaholab/moose#19525] [Q&A General] [answered]
Hello,
I am doing modification with the PenaltyInterfaceDiffusion kernel to achieve a function making the flux on the interface of two domains equals to exp(u-u_neighbor). However, the solver cannot reach convergence. I don't know if PenaltyInterface is the right way to do this or I write something wrong within the code. 

Here is an illustration of the model:
![图片](https://user-images.githubusercontent.com/95432912/144477452-86ae18e5-27af-45d3-bc25-597ebedbd807.png)

Here is my repository for the files:

https://github.com/Chaoyue-Liu/test.git

Any suggestions will be appreciated! Thank you.
Chao

**Comment by jiangwen84** [comment, marked as answer]
I suggest you write your own interface kernel deriving from `InterfaceKernel` without modifying the existing `PenaltyInterfaceDiffusion`. `PenaltyInterfaceDiffusion` is used to enforce the equal (jump) value constraint on the variables. In your new class, the residual on element and neighbor will be something like : 

```
r = _test[_i][_qp] * (exp(_u[_qp] - _neighbor_value[_qp]));
```

and 

```
r = _test_neighbor[_i][_qp] * (exp(_u[_qp] - _neighbor_value[_qp]));
```

Double check the sign in those residual. I am not quite understand your definition of `normal`, e.g. the convention for the direction definition.

You need to derive the Jacobian as well, e.g., take the derivative of `r` respect to `_u` and `_neighbor_value`.

**Reply by Chaoyue-Liu** [reply]
Thanks jiangwen for the prompt reply. I checked my residual and Jacobian again, I think it's fine which I post below, in case I did something wrong:

my RESIDUAL:
jump_value =  exp(_u[_qp] - _neighbor_value[_qp]);
  switch (type)
  {
    case Moose::Element:
      r = _test[_i][_qp] * _penalty * jump_value;
      break;

    case Moose::Neighbor:
      r = _test_neighbor[_i][_qp] * -_penalty * jump_value ;
      break;
  }

my JACOBIAN:
  switch (type)
  {
    case Moose::ElementElement:
      jac = _test[_i][_qp] * _penalty * _phi[_j][_qp] * exp(_u[_qp] - _neighbor_value[_qp]);
      break;

    case Moose::ElementNeighbor:
      jac = _test[_i][_qp] * _penalty * -_phi_neighbor[_j][_qp] * exp(_u[_qp] - _neighbor_value[_qp]);
      break;

    case Moose::NeighborElement:
      jac = _test_neighbor[_i][_qp] * -_penalty * _phi[_j][_qp] * exp(_u[_qp] - _neighbor_value[_qp]);
      break;

    case Moose::NeighborNeighbor:
      jac = _test_neighbor[_i][_qp] * -_penalty * -_phi_neighbor[_j][_qp] * exp(_u[_qp] - _neighbor_value[_qp]);
      break;
  }

Also I tested other formula e.g. _u[_qp]*_u[_qp}-_u_neighbor[_qp], which works fine. But when it comes to exponential or sin or square function, it becomes non convergence.

**Reply by jiangwen84** [reply]
1. use penalty = 1.
2. apply the chain rule in your Jacobian calculation. For example, d(exp(f))/dx = exp(f) df/dx.

**Reply by Chaoyue-Liu** [reply]
> 1. use penalty = 1.
> 
>     2. apply the chain rule in your Jacobian calculation. For example, d(exp(f))/dx = exp(f) df/dx.

yes, that's what I did. It just don't converge somehow.

**Reply by Chaoyue-Liu** [reply]
I checked the model again. Turns out I made a silly mistake. The code modification is fine, instead, the sign of my neumann boundary condition is wrong. Thanks for the help!
